File: src/lib/scraper/utils/proxy-manager.ts
/**
 * Proxy Manager dla scrapera
 * 
 * Klasa do zarządzania i rotacji proxy
 */

interface Proxy {
  host: string;
  port: number;
  auth?: {
    username: string;
    password: string;
  };
  blockedUntil?: number;
}

export class ProxyManager {
  private proxies: Proxy[];
  private currentIndex: number;
  private blockDuration: number; // Czas blokady proxy w milisekundach
  
  /**
   * Konstruktor
   * @param proxyList Lista proxy w formacie ['http://user:pass@host:port']
   */
  constructor(proxyList: string[] = []) {
    this.proxies = this.parseProxyList(proxyList);
    this.currentIndex = 0;
    this.blockDuration = 5 * 60 * 1000; // 5 minut
  }
  
  /**
   * Parsowanie listy proxy z formatu string do obiektu Proxy
   */
  private parseProxyList(proxyList: string[]): Proxy[] {
    return proxyList.map(proxyStr => {
      try {
        const url = new URL(proxyStr);
        const auth = url.username && url.password ? {
          username: url.username,
          password: url.password
        } : undefined;
        
        return {
          host: url.hostname,
          port: parseInt(url.port, 10) || 80,
          auth
        };
      } catch (error) {
        console.error(`Invalid proxy format: ${proxyStr}`);
        return null;
      }
    }).filter(Boolean) as Proxy[];
  }
  
  /**
   * Pobieranie następnego dostępnego proxy
   */
  getNextProxy(): Proxy | null {
    if (this.proxies.length === 0) {
      return null;
    }
    
    const now = Date.now();
    let checkedCount = 0;
    
    while (checkedCount < this.proxies.length) {
      const proxy = this.proxies[this.currentIndex];
      this.currentIndex = (this.currentIndex + 1) % this.proxies.length;
      
      if (!proxy.blockedUntil || proxy.blockedUntil < now) {
        return proxy;
      }
      
      checkedCount++;
    }
    
    // Jeśli wszystkie proxy są zablokowane, zwróć null
    return null;
  }
  
  /**
   * Oznaczenie proxy jako zablokowanego
   */
  markAsBlocked(proxy: Proxy | null): void {
    if (!proxy) return;
    
    const index = this.proxies.findIndex(p => 
      p.host === proxy.host && p.port === proxy.port
    );
    
    if (index !== -1) {
      this.proxies[index].blockedUntil = Date.now() + this.blockDuration;
    }
  }
}

File: school-finder-scraper/utils/proxy_manager.py
import random
import time

class ProxyManager:
    """
    Klasa do zarządzania listą proxy i rotacji między nimi
    """
    def __init__(self, proxies=None):
        """
        Inicjalizacja menedżera proxy
        
        Args:
            proxies (list, optional): Lista proxy w formacie {'http': 'http://user:pass@host:port'}
        """
        self.proxies = proxies or []
        self.blocked_proxies = {}
        self.current_index = 0
    
    def get_next_proxy(self):
        """
        Pobieranie następnego dostępnego proxy z listy
        
        Returns:
            dict: Proxy w formacie {'http': 'http://user:pass@host:port'} lub pusty słownik
        """
        # Jeśli lista proxy jest pusta, zwracamy pusty słownik
        if not self.proxies:
            return {}
        
        # Sprawdzenie, czy są dostępne proxy (nie zablokowane)
        available_proxies = [p for p in self.proxies if p not in self.blocked_proxies or 
                            time.time() - self.blocked_proxies[p] > 300]  # 5 minut blokady
        
        if not available_proxies:
            # Jeśli wszystkie proxy są zablokowane, resetujemy blokady dla najstarszych
            if self.blocked_proxies:
                oldest_blocked = min(self.blocked_proxies.items(), key=lambda x: x[1])[0]
                del self.blocked_proxies[oldest_blocked]
                available_proxies = [oldest_blocked]
            else:
                return {}
        
        # Wybieramy losowe proxy z dostępnych
        proxy = random.choice(available_proxies)
        return proxy
    
    def mark_as_blocked(self, proxy):
        """
        Oznaczenie proxy jako zablokowanego
        
        Args:
            proxy (dict): Proxy do oznaczenia jako zablokowane
        """
        if proxy and proxy in self.proxies:
            self.blocked_proxies[proxy] = time.time()
    
    def add_proxy(self, proxy):
        """
        Dodanie nowego proxy do listy
        
        Args:
            proxy (dict): Proxy do dodania
        """
        if proxy and proxy not in self.proxies:
            self.proxies.append(proxy)
    
    def remove_proxy(self, proxy):
        """
        Usunięcie proxy z listy
        
        Args:
            proxy (dict): Proxy do usunięcia
        """
        if proxy in self.proxies:
            self.proxies.remove(proxy)
            if proxy in self.blocked_proxies:
                del self.blocked_proxies[proxy]